---
title: "$F_{ST}$ within lakes."
author: "J. Ignacio Lucas LledÃ³ & Mar Llaberia-Robledillo"
date: "`r Sys.Date()`"
output:
   html_document:
      toc: TRUE
---

Species in the same lake are supposed to hybridize occasionally, so that they may
share variation along large portions of the genome. That is: the $F_{ST}$ between species
in the same lake are expected to be low, in general. But loci that contribute to the
adaptation of an ecomorph must remain divergent between ecomorphs in the same lake.
In a background of low $F_{ST}$ values, it should be easy to spot the positions with
high $F_{ST}$ values.

# Alpine region 
I will compare first *C. fatioi* (Felchen ecomorph, Thun and Brienz lakes) with
*C. steinmanni* (Balchen ecomorph, in Thun). 

```{bash Thun-Brienz}
VCF=../2022-04-22/assem2_outfiles/assem2.vcf
FISHDATA=../../data/Fish_clean2.tsv

if [ ! -d FST ]; then mkdir FST; fi

if [ ! -e FST/fatioi-steinmanni.windowed.weir.fst ]; then
   if [ ! -e dirty.recode.vcf ]; then
      vcftools --vcf $VCF \
               --out dirty \
               --remove ../2022-05-04/TheWorst.txt \
               --recode \
               --recode-INFO-all > vcftools.log 2>&1 
   fi

   if [ ! -e C.fatioi.txt ]; then
      gawk '($19 == "C.fatioi"){print $1}' $FISHDATA > C.fatioi.txt
   fi

   if [ ! -e C.steinmanni ]; then
      gawk '($19 == "C.steinmanni"){print $1}' $FISHDATA > C.steinmanni.txt
   fi

   vcftools --vcf dirty.recode.vcf \
            --out FST/fatioi-steinmanni \
            --thin 500 \
            --remove-indels \
            --mac 2 \
            --max-alleles 2 \
            --max-missing 1 \
            --max-meanDP 1000 \
            --weir-fst-pop C.fatioi.txt \
            --weir-fst-pop C.steinmanni.txt \
            --fst-window-size 1000000 \
            --fst-window-step  100000 >> vcftools.log 2>&1
fi
```

```{r plots1}
suppressMessages(library(Biostrings))
suppressMessages(library(GenomicRanges))
suppressMessages(library(ggplot2))
ref             <- readDNAStringSet('../../data/reference.fa')
# The names of chromosomes must be only the first word in the fasta name line.
names(ref)      <- sapply(strsplit(names(ref), " "), '[[', 1)
# the refWidth table contains the lengths of chromosomes.
refWidth        <- width(ref)
names(refWidth) <- names(ref)

fst1 <- read.table('FST/fatioi-steinmanni.windowed.weir.fst',
                  header = TRUE, stringsAsFactors = FALSE)
fst1$POS        <- (fst1$BIN_END + fst1$BIN_START) / 2
offset          <- cumsum(refWidth)[1:39]
names(offset)   <- names(ref)[2:40]
offset['LR778253.1'] <- 0
fst1$offset      <- offset[fst1$CHROM]
fst1$POS2        <- fst1$POS + fst1$offset
ChromCol        <- rep(c(1,2), 20)
names(ChromCol) <- names(ref)
fst1$ChromCol    <- ChromCol[fst1$CHROM]
ggplot(data = fst1, mapping = aes(x = POS2, y = WEIGHTED_FST, color = ChromCol)) +
   geom_point() + theme(legend.position="none")
```

The equivalent comparison between a Felchen and a Balchen ecomorph in a different
lake is the comparsion of *C. zuerichensis* (Felchen) and *C. duplex* (Balchen),
both from the Zurich and Walen lakes. If natural selection targeted the same loci
in both lakes to produce the adaptations that define the Felchen and Balchen
ecomorphs, then, several windows with a high $F_{ST}$ above should also have a
high $F_{ST}$ in the results below. Let's see.

```{bash zuerichensis-duplex}
FISHDATA=../../data/Fish_clean2.tsv

if [ ! -e FST/zuerichensis-duplex.windowed.weir.fst ]; then
   if [ ! -e C.zuerichensis.txt ]; then
      gawk '($19 == "C.zuerichensis"){print $1}' $FISHDATA > C.zuerichensis.txt
   fi
   
   if [ ! -e C.duplex.txt ]; then
      gawk '($19 == "C.duplex"){print $1}' $FISHDATA > C.duplex.txt
   fi
   vcftools --vcf dirty.recode.vcf \
            --out FST/zuerichensis-duplex \
            --thin 500 \
            --remove-indels \
            --mac 2 \
            --max-alleles 2 \
            --max-missing 1 \
            --max-meanDP 1000 \
            --weir-fst-pop C.zuerichensis.txt \
            --weir-fst-pop C.duplex.txt \
            --fst-window-size 1000000 \
            --fst-window-step  100000 >> vcftools.log 2>&1
fi
```


```{r plots2}
fst2 <- read.table('FST/zuerichensis-duplex.windowed.weir.fst',
                   header = TRUE, stringsAsFactors = FALSE)
fst2$POS        <- (fst2$BIN_END + fst2$BIN_START) / 2
offset          <- cumsum(refWidth)[1:39]
names(offset)   <- names(ref)[2:40]
offset['LR778253.1'] <- 0
fst2$offset     <- offset[fst2$CHROM]
fst2$POS2       <- fst2$POS + fst2$offset
ChromCol        <- rep(c(1,2), 20)
names(ChromCol) <- names(ref)
fst2$ChromCol   <- ChromCol[fst2$CHROM]
ggplot(data = fst2, mapping = aes(x = POS2, y = WEIGHTED_FST, color = ChromCol)) +
   geom_point() + theme(legend.position="none")
```

```{r comparacio1}
# Thun-Brienz:
thrs1 <- quantile(fst1$WEIGHTED_FST, probs = 0.995)
f1 <- fst1$WEIGHTED_FST >= thrs1
# Walen-Zurich:
thrs2 <- quantile(fst2$WEIGHTED_FST, probs = 0.995)
f2 <- fst2$WEIGHTED_FST >= thrs2

GR1 <- GRanges(
   seqnames = factor(fst1[f1, 'CHROM']),
   ranges = IRanges(start = fst1[f1, 'BIN_START'],
                    end = fst1[f1, 'BIN_END'],
                    names = paste0(fst1[f1, 'CHROM'], ':',
                                   fst1[f1, 'BIN_START'], '-',
                                   fst1[f1, 'BIN_END'],
                                   sep = '')),
   strand = '+',
   seqinfo = Seqinfo(seqnames = names(ref),
                     seqlengths = width(ref))
)
GR1 <- reduce(GR1)

GR2 <- GRanges(
   seqnames = factor(fst2[f2, 'CHROM']),
   ranges = IRanges(start = fst2[f2, 'BIN_START'],
                    end = fst2[f2, 'BIN_END'],
                    names = paste0(fst2[f2, 'CHROM'], ':',
                                   fst2[f2, 'BIN_START'], '-',
                                   fst2[f2, 'BIN_END'],
                                   sep = '')),
   strand = '+',
   seqinfo = Seqinfo(seqnames = names(ref),
                     seqlengths = width(ref))
)
GR2 <- reduce(GR2)

findOverlaps(GR1, GR2)

GR <- data.frame(
   chr = c(seqnames(GR1), seqnames(GR2)),
   start = c(start(GR1), start(GR2)),
   end = c(end(GR1), end(GR2)),
   lake = c(rep('Thun-Brienz', length(GR1)),
            rep('Walen-Zurich', length(GR2))),
   height = c(rep(1, length(GR1)),
              rep(2, length(GR2)))
)

for (chr in unique(GR$chr)) {
   f <- GR$chr == chr
   p <- ggplot(data = GR[f,],
               mapping = aes(x = start,
                             y = height,
                             xend = end,
                             yend = height,
                             color = lake)) +
      geom_segment(size = 5) +
      ylim(0,3) +
      xlim(0, width(ref[chr])) +
      theme(aspect.ratio = 0.3,
            axis.text.y = element_blank(),
            axis.line.y = element_blank(),
            axis.title.y = element_blank(),
            axis.ticks.y = element_blank()) +
      ggtitle(chr) + xlab('Position (bp)')
   print(p)
}

```


# Arctic region 
First, I will compare P and L ecomorphs of lake SUO. I will recover of the text files created in the previous folder (2022-12-23) for the estimation.

```{bash P-L-SUO}

if [ ! -e FST/P-L-SUO.windowed.weir.fst ]; then
   P=../2022-12-23/SUO_P.txt
   L=../2022-12-23/SUO_L.txt
   
   vcftools --vcf dirty.recode.vcf \
            --out FST/P-L-SUO \
            --thin 500 \
            --remove-indels \
            --mac 2 \
            --max-alleles 2 \
            --max-missing 1 \
            --max-meanDP 1000 \
            --weir-fst-pop $P \
            --weir-fst-pop $L \
            --fst-window-size 1000000 \
            --fst-window-step  100000
fi
```


```{r plots3}
fst <- read.table('FST/P-L-SUO.windowed.weir.fst',
                  header = TRUE, stringsAsFactors = FALSE)
fst$POS    <- (fst$BIN_END + fst$BIN_START) / 2
offset     <- cumsum(refWidth)[1:39]
names(offset) <- names(ref)[2:40]
offset['LR778253.1'] <- 0
fst$offset <- offset[fst$CHROM]
fst$POS2 <- fst$POS + fst$offset
ChromCol <- rep(c(1,2), 20)
names(ChromCol) <- names(ref)
fst$ChromCol <- ChromCol[fst$CHROM]
ggplot(data = fst, mapping = aes(x = POS2, y = WEIGHTED_FST, color = ChromCol)) +
   geom_point() + theme(legend.position="none")
```

Second, I'm goingo to compare P and L of lake LAN. The same as above. 

```{bash P-L-LAN}

if [ ! -e FST/P-L-LAN.windowed.weir.fst ]; then
   P=../2022-12-23/LAN_P.txt
   L=../2022-12-23/LAN_L.txt
   
   vcftools --vcf dirty.recode.vcf \
            --out FST/P-L-LAN \
            --thin 500 \
            --remove-indels \
            --mac 2 \
            --max-alleles 2 \
            --max-missing 1 \
            --max-meanDP 1000 \
            --weir-fst-pop $P \
            --weir-fst-pop $L \
            --fst-window-size 1000000 \
            --fst-window-step  100000
fi
```


```{r plots4}
fst <- read.table('FST/P-L-LAN.windowed.weir.fst',
                  header = TRUE, stringsAsFactors = FALSE)
fst$POS    <- (fst$BIN_END + fst$BIN_START) / 2
offset     <- cumsum(refWidth)[1:39]
names(offset) <- names(ref)[2:40]
offset['LR778253.1'] <- 0
fst$offset <- offset[fst$CHROM]
fst$POS2 <- fst$POS + fst$offset
ChromCol <- rep(c(1,2), 20)
names(ChromCol) <- names(ref)
fst$ChromCol <- ChromCol[fst$CHROM]
ggplot(data = fst, mapping = aes(x = POS2, y = WEIGHTED_FST, color = ChromCol)) +
   geom_point() + theme(legend.position="none")
```
