---
title: "vcftools analysis"
author: "Mar Llaberia-Robledillo"
date: "4/5/2022"
output: html_document
---

With assembly to the reference genome and clustering at 93\% similarity complete, the output is a .vcf file. We are going to work with the vcftools program to summarize, filter and run some calcuations to our genetic variation data. 

# 1. Individual Filtering Options 

We know individual SUO095 does not have enough data. But there may be others. Before
deciding what individuals to remove from the data set, we need to see the proportion
of sites where they miss data.

```{bash missingness}
VCF='../2022-04-22/assem2_outfiles/assem2.vcf'
if [ ! -e assem2.imiss ]; then
   vcftools --vcf $VCF \
            --out assem2 \
            --missing-indv
fi
```

```{r missing}
missingness <- read.table('assem2.imiss', header = TRUE)
plot(ecdf(missingness$F_MISS),
     xlab = 'Fraction of sites missing',
     main = 'ECDF of fraction of sites missing per individual')
abline(v = 0.82, col = 'red', lty = 2)
abline(v = 0.655, col = 'orange', lty = 2)
VeryBad <- as.character(missingness[missingness$F_MISS > 0.83, 'INDV'])
VeryBad
QuiteBad <- as.character(missingness[missingness$F_MISS > 0.655 & missingness$F_MISS <= 0.83, 'INDV'])
QuiteBad
if (! file.exists('VeryBad.txt')) {
  write.table(VeryBad, file = 'VeryBad.txt', quote = FALSE, row.names = FALSE, col.names = FALSE)
}
```

The raw data includes 879238 sites, about 40% of which are missing in any individual
with good coverage. There is a tail of individuals with low coverage that miss more
than 50% of sites. Many of them are usable in a large enough dataset, but this result
makes me think that we should build two data sets: one to optimize the number of individuals
and one to optimize the number of sites. However, the fact that 80% of individuals miss
less than 60% of sites does not mean that the sites they share are 60% of the total:
every individual may miss a different subset of the sites.

```{bash}
VCF='../2022-04-22/assem2_outfiles/assem2.vcf'
if [ ! -e assem2_indv.recode.vcf ]; then
   vcftools --vcf $VCF \
            --out assem2_indv \
            --remove VeryBad.txt \
            --recode \
            --recode-INFO-all
fi
```

# 2. Site Filtering Options

### Position filtering 
-- thin Specified distance between one site and the next.

### Variant type Filtering 
-- remove-indels Exclude sites that contain an idel (any variant that alters the lenght of the REF allele). 

### Info Field Filtering 
-- mac Include only sites with Minor Allele Count higher or equal to <value>
-- max-allels Include only sites with a number of alleles lower or equal to <value>

### Genotype Value Filtering 
--max-meanDP Includes only sites with mean depth values (over all included individuals) lower or equal to the <value>. These options require that the "DP" FORMAT tag is included for each site.
--max-missing count Exclude sites with more than this number of missing genotypes over all individuals.

### VCF Output Options 
--recode Generate a new .vcf file from the input after applying the filtering options specified. 
--recode-INFO-all Keep the INFO key present in the input file.

```{bash}
if [ ! -e assem2_filter.recode.vcf ]; then
   vcftools --vcf assem2_indv.recode.vcf \
            --out assem2_filter \
            --thin 500 \
            --remove-indels \
            --mac 2 \
            --max-alleles 2 \
            --max-meanDP 1000 \
            --max-missing-count 12 \
            --recode --recode-INFO-all
fi
```

# 3. Output Options 

### LD Statistics 
--hap-r2 Outputs a file reporting the r2, D, and Dâ€™ statistics using phased haplotypes. These are the traditional measures of LD often reported in the population genetics literature.

### Other Statistics 

--relatedness This option is used to calculate and output a relatedness statistic based on the method of Yang et al. 2010. Expectation of Ajk is zero for individuals within a populations, and one for an individual with themselves (unadjusted Ajk statistic). The output file has the suffix ".relatedness".
--relatedness2 This option is used to calculate and output a relatedness statistic based on the method of Manichaikul et al.2010. The output file has the suffix ".relatedness2".
--missing-indv Generates a file reporting the missingness on a per-individual basis. The file has the suffix ".imiss".

In this case, it is not necessary to create an output file (option --recode), the program creates it for each statistic.

The relatedness would make more sense to calculate it only for members of the same species. We
skip that for the moment.

```{bash}
INPUT='assem2_filter.recode.vcf'
OUTPUT='assem2_outstat'
#if [ ! -e assem2_outstat.relatedness ]; then
#   vcftools --vcf $INPUT --out $OUTPUT --relatedness
#fi
#if [ ! -e assem2_outstat.relatedness2 ]; then
#   vcftools --vcf $INPUT --out $OUTPUT --relatedness2
#fi
if [ ! -e assem2_outstat.imiss ]; then
   vcftools --vcf $INPUT --out $OUTPUT --missing-indv
fi
```

```{r again}
imiss <- read.table('assem2_outstat.imiss', header = TRUE, row.names = 1)
plot(ecdf(imiss$F_MISS), xlab = 'Fraction of missing sites per individual',
     main = 'ECDF of fraction of missing sites per individual in filtered VCF')
abline(v = 0.10, col = 'red', lty = 2)
NotGood <- row.names(imiss[imiss$F_MISS > 0.10,])
NotGood
if (! file.exists('NotGood.txt')) {
  write.table(NotGood, file = 'NotGood.txt', quote = FALSE,
              row.names = FALSE, col.names = FALSE)
}
```

There are still 4 individuals with more than 10% of sites missing. These are the same
4 individuals that would have been removed next in the individual filter above, if we
had be a bit more strict. If we do remove them before filtering sites, we would probably
obtan a more complete matrix of genotypes. Let's see.

## More strict filtering

```{bash}
VCF='../2022-04-22/assem2_outfiles/assem2.vcf'
if [ ! -e TheWorst.txt ]; then
   cat VeryBad.txt NotGood.txt > TheWorst.txt
fi
if [ ! -e assem2_indv2.recode.vcf ]; then
   vcftools --vcf $VCF \
            --out assem2_indv2 \
            --remove TheWorst.txt \
            --recode \
            --recode-INFO-all
fi

# To make sure site filters are applied after individual filters, we now run
# the following separately:

if [ ! -e strict.recode.vcf ]; then
   vcftools --vcf assem2_indv2.recode.vcf \
            --out strict \
            --thin 500 \
            --remove-indels \
            --mac 2 \
            --max-alleles 2 \
            --max-meanDP 1000 \
            --max-missing-count 12 \
            --recode --recode-INFO-all
fi

if [ ! -e assem2_strict.imiss ]; then
   vcftools --vcf strict.recode.vcf \
            --out assem2_strict \
            --missing-indv
fi
```


```{r last}
imiss2 <- read.table('assem2_strict.imiss', header = TRUE, row.names = 1)
plot(ecdf(imiss2$F_MISS), xlab = 'Proportion of sites missing per individual',
     main = 'ECDF of proportion of missing sites per individual')
f <- imiss2$F_MISS > 0.05
z <- imiss2[f,]
z$INDV <- row.names(z)
z <- z[order(z$F_MISS),]
text(z$F_MISS, labels = z$INDV, y = c(0.8,0.85,0.9,0.7,0.75,0.8,0.9,0.9), adj=0)
#row.names(imiss2[imiss2$F_MISS > 0.1,])
```

There are still 2 individuals with more than 10% of missing data. But now we have
24878 sites. I think we can live with this. If individuals with a higher fraction
of missing data happened to look like hybrids, I would repeat the analyses without
many of those missing sites.

```{r sessioninfo}
sessionInfo()
```

