---
title: "ipyrad analysis 0.93"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Branching

```{bash}

## Create a new branch of the Assembly "both"
#ipyrad -p params_both.txt -b assem2
## Move the new files into a new folder
# mv assem2.json params_assem2.txt ~/cophyhopa/results/2022-04-22
```

# Assembly analysis with 93% of similarity

Here, we continue with the Assembly with the new branching data. First, we change the parameters file `(params-assem2.txt`) according to the new values that interest us. The major change is that the clustering is now executed based on the 93% similarity.

The defined parameters (`params-assem2.txt`) for all Steps are:

-   `[0] [assembly_name]`: Used to name output directories for assembly steps.
-   `[1] [project_dir]`: Project directory (made in curdir if not present)
-   `[2] [raw_fastq_path]`: Location of raw non-demultiplexed `fastq`files (empty line)
-   `[3] [barcodes_path]`: Location of barcodes file (empty line)
-   `[4] [sorted_fastq_path]`: Location of demultiplexed/sorted `fastq`files
-   `[5] [assembly_method]`: Assembly method (`reference`)
-   `[6] [reference_sequence]`: Location of reference sequence file
-   `[7] [datatype]`: Datatype (`pairddrad`)
-   `[8] [restriction_overhang]`: Restriction overhang (empty line in our case)
-   `[9] [max_low_qual_bases]`: Max low quality base calls (Q\<20) in a read (set to `4`)
-   `[10] [phred_Qscore_offset]`: phred Q score offset (set to `33`)
-   `[11] [mindepth_statistical]`: Min depth for statistical base calling (set to `6`)
-   `[12] [mindepth_majrule]`: Min depth for majority-rule base calling (set to `6`)
-   `[13] [maxdepth]`: Max cluster depth within samples (set to `10000`)
-   `[14] [clust_threshold]`: Clustering threshold for de novo assembly (set to `0.93`)
-   `[15] [max_barcode_mismatch]`: Max number of allowable mismatches in barcodes (`0`)
-   `[16] [filter_adapters]`: Filter for adapters/primers (set to `2`)
-   `[17] [filter_min_trim_len]`: Min length of reads after adapter trim (set to `35`)
-   `[18] [max_alleles_consens]`: Max alleles per site in consensus sequences (set to `2`)
-   `[19] [max_Ns_consens]`: Max N's (uncalled bases) in consensus (set to `0.05`)
-   `[20] [max_Hs_consens]`: Max Hs (heterozygotes) in consensus (set to `0.5`)
-   `[21] [min_samples_locus]`: Min \# samples per locus for output (set to `4`)
-   `[22] [max_SNPs_locus]`: Max \# SNPs per locus (set to `0.1`)
-   `[23] [max_Indels_locus]`: Max \# of indels per locus (set to `2`)
-   `[24] [max_shared_Hs_locus]`: Max \# heterozygous sites per locus (set to `0.5`)
-   `[25] [trim_reads]`: Trim raw reads edges (`0, -10, 0, -10`)
-   `[26] [trim_loci]`: Trim locus edges (`0, 0, 0, 0`)
-   `[27] [output_formats]`: Output formats (set to `*` -all-)
-   `[28] [pop_assign_file]`: Path to population assignment file

```{bash}
conda activate ipyrad

## edit the params to enter its raw_fastq_path and barcodes file
#nano params-assem2.txt

ipyrad -p params-assem2.txt -s 34567 c- 24
ipyrad -p params-assem2.txt -r

```
