\documentclass[a4paper,12pt]{article}
\usepackage[utf8]{inputenc}
\usepackage[T1]{fontenc}
\usepackage{booktabs}
\usepackage{graphicx}
\usepackage{rotating}
\usepackage{color}
\usepackage{fancyvrb}
\usepackage[left=3cm]{geometry}
\usepackage{natbib}
\bibliographystyle{plainnat}
\author{J. Ignacio Lucas Lled√≥}
\title{ddRADseq library design}
\date{Last update: \today}
\begin{document}
\maketitle
\begin{flushleft}
This is based on \citet{Salas-Lizana2018}. \emph{SphI} recognizes the sequence \textsf{GCATGC} and produces \textsf{CATG}-3' overhangs. And \emph{HindIII} recognizes \textsf{AAGCTT} and produces 5'-\textsf{AGCT} overhangs. Below I show the design for the adapters.

\subsection*{\emph{HindIII}-P1 adaptor}
\begin{tabular}{l}
\verb+*AGCTCTGTCTCTTATACGAGAACAA+\\
\verb+     |||||||||||||+\\[-8pt]
   \begin{turn}{180}
   \verb+GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG     +
   \end{turn}
\\
\end{tabular}
\vspace*{0.3cm}

\subsection*{\emph{SphI}-P2 adaptor}

\begin{tabular}{l}
\verb+  GTCGGCAGCGTCAGATGTGTATAAGAGACAGCCATG+\\
\verb+  ||||||||||||||||||||||||||||||||+\\[-8pt]
   \begin{turn}{180}
   \verb+*GCTGTCTCTTATACACATCTGACGCTGCCGACGA+
   \end{turn}
\\
\end{tabular}
\vspace*{0.3cm}

\subsection*{Ligated fragment}
Note that once the adapters are ligated to the genomic fragment (blue), the restriction sites are not available any more, and the fragment would not be digested again. 

\fvset{fontsize=\scriptsize,commandchars=\\\{\}}
\begin{tabular}{l}
\Verb+  GTCGGCAGCGTCAGATGTGTATAAGAGACAGCCATG\textcolor{blue}{CNNNNNNNNNNA}AGCTCTGTCTCTTATACGAGAACAA+\\[-6pt]
\Verb+  ||||||||||||||||||||||||||||||||||||\textcolor{blue}{||||||||||||}|||||||||||||||||+\\[-10pt]
   \begin{turn}{180}
   \Verb+GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG\textcolor{blue}{AGCTTNNNNNNNNNNGCATG}GCTGTCTCTTATACACATCTGACGCTGCCGACGA+
   \end{turn}
\\
\end{tabular}
\vspace*{0.3cm}

\subsection*{Ligated fragment with amplification primers}
We use Nextera Index Kit v2 primers to incorporate unique combinations of adapters to every sample. Below, the positions marked as \emph{[i5]} and \emph{[i7]} correspond to the 10-bases long indices. Note these are not Unique Dual Indices, because every individual primer is shared with other samples, but the combination is sample-specific. In order to reduce the chance of index hopping, we include a clean-up step before and after the PCR amplification that incorporates the indices. In addition, we include we reserved two pairs of primers to label two samples with unique individual indices, which will allow us to quantify the rate of index hopping.

\fvset{fontsize=\tiny}
\begin{tabular}{l}
\Verb+AATGATACGGCGACCACCGAGATCTACAC[i5]TCGTCGGCAGCGTC+\\[-8pt]
\Verb+                                   GTCGGCAGCGTCAGATGTGTATAAGAGACAGCCATG\textcolor{blue}{CNNNNNNNNNNA}AGCTCTGTCTCTTATACGAGAACAA+\\[-6pt]
\Verb+                                   ||||||||||||||||||||||||||||||||||||\textcolor{blue}{||||||||||||}|||||||||||||||||+\\[-10pt]
   \begin{turn}{180}
   \Verb+GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG\textcolor{blue}{AGCTTNNNNNNNNNNGCATG}GCTGTCTCTTATACACATCTGACGCTGCCGACGA                                 +
   \end{turn}
\\[-8pt]
   \begin{turn}{180}
   \Verb+CAAGCAGAAGACGGCATACGAGAT[i7]GTCTCGTGGGCTCGG                                                                                                          +
   \end{turn}
\\
\end{tabular}
\vspace*{0.3cm}

\subsection*{Amplified fragment}
After ligation of adapters and incorporation of indices by amplification PCR, the original DNA fragments will be almost 150 bases longer. In any case, because both enzymes are rare cutters, we will select a wide range of fragment sizes to sequence, say between 300 and 900.

\begin{tabular}{l}
\Verb+AATGATACGGCGACCACCGAGATCTACAC[i5]TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGCCATG\textcolor{blue}{CNNNNNNNNNNA}AGCTCTGTCTCTTATACACATCTCCGAGCCCACGAGAC[i7]ATCTCGTATGCCGTCTTCTGCTTG+\\[-8pt]
\Verb+||||||||||||||||||||||||||||| || ||||||||||||||||||||||||||||||||||||||\textcolor{blue}{||||||||||||}|||||||||||||||||||||||||||||||||||||| || ||||||||||||||||||||||||+\\[-10pt]
\begin{turn}{180}
\Verb+CAAGCAGAAGACGGCATACGAGAT[i7]GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG\textcolor{blue}{AGCTTNNNNNNNNNNGCATG}GCTGTCTCTTATACACATCTGACGCTGCCGACGA[i5]GTGTAGATCTCGGTGGTCGCCGTATCATT+
\end{turn}
\\
\end{tabular}

\subsection*{Sequencing primers}
The sequencing primers only need to be edited to match the restriction site. This makes the sequence read start at the unknown genomic base, to optimize the yield and increase the complexity of the signal.

\begin{tabular}{l}
\Verb+                                 TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGCCATGC+\\[-8pt]
\Verb+AATGATACGGCGACCACCGAGATCTACAC[i5]TCGTCGGCAGCGTCAGATGTGTATAAGAGACAGCCATG\textcolor{blue}{CNNNNNNNNNNA}AGCTCTGTCTCTTATACACATCTCCGAGCCCACGAGAC[i7]ATCTCGTATGCCGTCTTCTGCTTG+\\[-8pt]
\Verb+||||||||||||||||||||||||||||| || ||||||||||||||||||||||||||||||||||||||\textcolor{blue}{||||||||||||}|||||||||||||||||||||||||||||||||||||| || ||||||||||||||||||||||||+\\[-10pt]
\begin{turn}{180}
   \Verb+CAAGCAGAAGACGGCATACGAGAT[i7]GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAG\textcolor{blue}{AGCTTNNNNNNNNNNGCATG}GCTGTCTCTTATACACATCTGACGCTGCCGACGA[i5]GTGTAGATCTCGGTGGTCGCCGTATCATT+
\end{turn}
\\[-8pt]
\begin{turn}{180}
   \Verb+GTCTCGTGGGCTCGGAGATGTGTATAAGAGACAGAGCTT                                                                                  +
\end{turn}
\\
\end{tabular}



\end{flushleft}
\bibliography{references} 
\end{document}