---
title: "Admixture analysis"
author: "Mar Llaberia-Robledillo"
date: "10/5/2022"
output: html_document
---

# 1. Convert .vcf file to PLINK format 

Admixture requires input data in PLINK bed format. So, we convert our .vcf file output to PLINK format, that includes .bed, .ped, .fam and .bim formats. 

```{bash}
VCF='/gata/mar/cophyhopa/results/2022-05-04/assem2_indv2.recode.vcf'
if [ ! -e ADMIX_ALL.vcf.gz ]; then
   vcftools --vcf $VCF \
            --out dirty \
            --recode \
            --recode-INFO-all
   vcftools --vcf dirty.recode.vcf \
            --out ADMIX_ALL \
            --thin 500 \
            --remove-indels \
            --mac 2 \
            --max-alleles 2 \
            --max-meanDP 1000 \
            --max-missing-count 7 \
            --stdout \
            --recode --recode-INFO-all \
            | gzip > ADMIX_ALL.vcf.gz
fi
if [ -e dirty.recode.vcf ]; then rm dirty.recode.vcf; fi

if [ ! -e chromMap.txt ]; then
   gunzip -c ADMIX_ALL.vcf.gz | \
   gawk '(/^[^#]/){
      A[$1] = 1
   }END{
      N = 1
      for (chr in A) {
         print chr "\t" N
         N = N + 1
      }
   }' > chromMap.txt
fi

gawk -v OFS="\t" '{
   $1 = 1
   $2 = 1 ":" $4
   print $0
}' ADMIX_ALL.map > better.map
mv better.map ADMIX_ALL.map

vcftools --gzvcf ADMIX_ALL.vcf.gz \
         --plink --mac 2 --remove-indels --max-alleles 2 \
         --chrom-map chromMap.txt \
         --out ADMIX_ALL

plink --file ADMIX_ALL --make-bed --out ADMIX_ALL --allow-no-sex --allow-extra-chr


```

# 2.  Admixture analysis with variant K 

For run admixture, we need to define the K parameter. Here, $K is a number indicating the number of clusters (populations) we want to infer. Since we don't know exactly how many clusters to define, we run the analysis with several number of clusters K (from 2 to 8), computing the cross-validation error. Then, we extract these values. 

```{bash}

INPUT='/gata/mar/cophyhopa/results/2022-05-09/admixture/ADMIX_ALL.bed'

for K in 7 8 9 10 11 12; do admixture --cv $INPUT $K | tee log${K}.out; done

```

```{bash echo = TRUE}
grep -h CV log*.out

```
# 3.  Plot Q estimates K=7

We choose the K that has given a lower value of the cross-validation error. We add the sample names and display a barplot 

```{r figure.align = "center", fig.height=6, fig.width=9}

library(RColorBrewer)
Q <- read.table("ADMIX_ALL.7.Q")
popmap <- read.table('~/cophyhopa/data/popmap.txt', 
                       col.names = c('sample', 'group', 'pop', 'lake'))

sample.id <- scan("ADMIX_ALL.nosex", what = "character")
sample.id <- sample.id[!duplicated(sample.id)]
rownames(popmap) <- popmap[,1]
popmap <- popmap[sample.id, ]

rownames(Q) <- sample.id
Q[,8] <- NA
Q[,8] <- popmap[rownames(Q), 3]

z <- order(round(Q[, 'V1'], 2),
           round(Q[, 'V2'], 2),
           round(Q[, 'V3'], 2),
           round(Q[, 'V4'], 2),
           round(Q[, 'V5'], 2),
           round(Q[, 'V6'], 2), decreasing = TRUE)

par(mar=c(7,4,4,2))
barplot(t(Q[z,]), col = brewer.pal(7, 'Accent'), border = NA,
        xlab = '', ylab = 'Admixture proportions for K=7', las=2,
        cex.names = 0.5, 
        names.arg = t(Q[z,8]))

```

# 4.  Plot Q estimates K=8

We choose the K that has given a lower value of the cross-validation error. We add the sample names and display a barplot 

```{r figure.align = "center", fig.height=6, fig.width=9}

library(RColorBrewer)
Q <- read.table("ADMIX_ALL.8.Q")
popmap <- read.table('~/cophyhopa/data/popmap.txt', 
                       col.names = c('sample', 'group', 'pop', 'lake'))

sample.id <- scan("ADMIX_ALL.nosex", what = "character")
sample.id <- sample.id[!duplicated(sample.id)]
rownames(popmap) <- popmap[,1]
popmap <- popmap[sample.id, ]

rownames(Q) <- sample.id
Q[,9] <- NA
Q[,9] <- popmap[rownames(Q), 3]

z <- order(round(Q[, 'V1'], 2),
           round(Q[, 'V2'], 2),
           round(Q[, 'V3'], 2),
           round(Q[, 'V4'], 2),
           round(Q[, 'V5'], 2),
           round(Q[, 'V6'], 2),
           round(Q[, 'V7'], 2), decreasing = TRUE)

par(mar=c(7,4,4,2))
barplot(t(Q[z,]), col = brewer.pal(8, 'Accent'), border = NA,
        xlab = '', ylab = 'Admixture proportions for K=8', las=2,
        cex.names = 0.5, 
        names.arg = t(Q[z,9]))

```

# 5.  Plot Q estimates K=9

We choose the K that has given a lower value of the cross-validation error. We add the sample names and display a barplot 

```{r figure.align = "center", fig.height=6, fig.width=9}

library(RColorBrewer)
Q <- read.table("ADMIX_ALL.9.Q")
popmap <- read.table('~/cophyhopa/data/popmap.txt', 
                       col.names = c('sample', 'group', 'pop', 'lake'))

sample.id <- scan("ADMIX_ALL.nosex", what = "character")
sample.id <- sample.id[!duplicated(sample.id)]
rownames(popmap) <- popmap[,1]
popmap <- popmap[sample.id, ]

rownames(Q) <- sample.id
Q[,10] <- NA
Q[,10] <- popmap[rownames(Q), 3]

z <- order(round(Q[, 'V1'], 2),
           round(Q[, 'V2'], 2),
           round(Q[, 'V3'], 2),
           round(Q[, 'V4'], 2),
           round(Q[, 'V5'], 2),
           round(Q[, 'V6'], 2),
           round(Q[, 'V7'], 2), decreasing = TRUE)

par(mar=c(7,4,4,2))
barplot(t(Q[z,]), col = brewer.pal(9, 'Accent'), border = NA,
        xlab = '', ylab = 'Admixture proportions for K=9', las=2,
        cex.names = 0.5, 
        names.arg = t(Q[z,10]))

```

# 6.  Plot Q estimates K=10

We choose the K that has given a lower value of the cross-validation error. We add the sample names and display a barplot 

```{r figure.align = "center", fig.height=6, fig.width=9}

library(RColorBrewer)
Q <- read.table("ADMIX_ALL.10.Q")
popmap <- read.table('~/cophyhopa/data/popmap.txt', 
                       col.names = c('sample', 'group', 'pop', 'lake'))

sample.id <- scan("ADMIX_ALL.nosex", what = "character")
sample.id <- sample.id[!duplicated(sample.id)]
rownames(popmap) <- popmap[,1]
popmap <- popmap[sample.id, ]

rownames(Q) <- sample.id
Q[,11] <- NA
Q[,11] <- popmap[rownames(Q), 3]

z <- order(round(Q[, 'V1'], 2),
           round(Q[, 'V2'], 2),
           round(Q[, 'V3'], 2),
           round(Q[, 'V4'], 2),
           round(Q[, 'V5'], 2),
           round(Q[, 'V6'], 2),
           round(Q[, 'V7'], 2), decreasing = TRUE)

par(mar=c(7,4,4,2))
barplot(t(Q[z,]), col = brewer.pal(10, 'Accent'), border = NA,
        xlab = '', ylab = 'Admixture proportions for K=8', las=2,
        cex.names = 0.5, 
        names.arg = t(Q[z,11]))

```

